Add AWS deploy guide to How To (#786)
meilisearch/documentation: create/how_to/aws.md

# Deploy a MeiliSearch Instance on AWS

Using our **MeiliSearch [AMI](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIs.html)**, MeiliSearch can be deployed on AWS in just a few minutes.
The following `how-to guide` will walk you through every step to deploy MeiliSearch in an AWS EC2 instance.

**Table of Contents**:

> [[toc]]

Reviewer: There is something weird, but it may not be just there, it may be on all the site. The permalink pointing to achors are working when you click them inside the page, but if you open a link on a new tab, it doesn't work. You can also see the issue if you copy a link pointing to an anchor, and paste it yourself in your browser, it will open the right page, but not scroll to the right position.

Author: I think this is a couple different behaviors.

1. If you select a header smaller than an `<h2>` the anchor link automatically changes to the parent header. I assume this is a VuePress thing but no idea if it's modifiable or not.
2.  If you open a page with lots of images (like our deploy guides), the page will scroll to the anchor BEFORE the images load. Then, when the images load in that pushes the content further down and you end up looking at the wrong place on the page.

Does that explain/cover the behavior you're seeing or is there another aspect to it? And do you see the same behavior on both the live site and the deploy previews?

Reviewer: No, it's a new one! Let me do a video :)

Reviewer: Or maybe it's the same issue, but check. The fact that VuePress cannot manage something as basic as links and anchors is another reason for switching IMNSHO

https://user-images.githubusercontent.com/1170989/108264838-06220e80-7136-11eb-961c-c88bc9ce948d.mp4

Reviewer: Anyway, it's a problem with VuePress, not this PR.

## Part 1: Deploy an out-of-the-box MeiliSearch instance

### 1. Launch an instance from the AWS console

After logging into your [AWS Console](https://aws.amazon.com/console), navigate to the **Compute** service. Then go to **EC2**, and finally open your **Instances** console.

> ![Launch an instance from AWS console](/aws/01.launch-instances.png)

Reviewer: The image quality isn't good. Not sure if it's the quality or the size. If it's the size that make it hard to read, we should link to the full size image. It's like this for all images.

Author: Are you sure it's a problem with the image itself? On my browser the deploy previews open very zoomed-out. A couple `⌘` + `+`s solves the problem.

Reviewer: So as we discovered, it's the size of the image that is causing problem, so I did another PR to add a zoom features. It doesn't change the fact that it's blurry, but it give the opportunity for people to see the full resolution version!

In the top-right corner, click on the **Launch instances** button to start the process of configuring your MeiliSearch instance.

### 2. Select 'MeiliSearch' AMI from 'Community AMIs'

You will now select which AMI or system Image to use to run your instance. Type **"meilisearch"** in the searchbar and select the **Community AMIs** tab on the left sidebar.

![Choose MeiliSearch AMI from Community AMIs](/aws/02.select-ami.png)

Click on **Select** (right side of the screen) to confirm your choice.

### 3. Size and specs

Select the specifications of the server you want MeiliSearch to run on.

![Size and specs](/aws/03.size-and-specs.png)

> > We recommend prioritizing memory allocation for better MeiliSearch performance.

Reviewer: I can't find where we mention this in the doc right now, but we should point to the section where we elaborate on minimum specifications or explanation on why memory is a possible issue. It will give more context for people that want it.

Author: Totally agree. It's [still a PR](https://github.com/meilisearch/documentation/pull/772)

Reviewer: oh, once @qdequele approve the PR, we can add the link. No need to wait for it for this PR, but be sure to remember to add it once it's merged.

> Note that the **free tier** is sufficient for tests or prototypes, but not recommended for large datasets.

Once you've made your choice, click on **Next: Configure instance details** to continue.

### 4. Instance details

> Here you can specify details of your MeiliSearch instance. Since **this section is not required to run MeiliSearch**, we won't cover it in this guide.

Reviewer: Should we point to the AWS doc talking about that part? Feel like asshole than just "we won't help you suckers" 🤣

> ![Instance details (optional)](/aws/04.instance-details.png)

Reviewer 2: ```suggestion
![Instance details](/aws/04.instance-details.png)
```

Simply click on **Next: Add Storage** to keep going.

### 5. Storage

Choose the storage **device** and **size** for your MeiliSearch instance.

![Storage](/aws/05.storage.png)

> The amount of storage space required can vary drastically depending on the data you plan to index. In this example, we will use 25 GiB, which is more than enough for a small to medium dataset.

Reviewer: small to medium dataset means nothing, can we be clearer

Author: It's a little arbitrary, but I would say that
Small dataset: less than 10,000 documents
Medium dataset: less than 1 million documents
Large dataset: 1 million documents and above
Does that sound right to you?

Author: But of course it's *all* arbitrary.

I'm sure you could easily pass 25gb with a dataset that has only 10,000 documents if those documents each had 1000+ fields of 1000 or more characters. There's no good way to escape the lack of clarity here, so I think maybe just change it to:

```suggestion
The amount of storage space required can vary drastically depending on the data you plan to index. In this example, we will use 25 GiB, which is more than enough for most small datasets.
```

![Domain to  MeiliSearch](/aws/11.domain.png)

Your domain name should now be linked to your MeiliSearch instance. Run a health check to verify that your instance is running and your DNS is well configured:

```bash
curl -v http://<your-domain-name>/health
```

The server should answer with a `204 No content` status code as shown in the example below:

```bash
...
< HTTP/1.1 204 No Content
...
```

### 2. Set API KEY and SSL (HTTPS)

Meilisearch is currently running in a *development environment*. You haven't set up an **API key**, meaning that anyone can read/write from your MeiliSearch, and you aren't using HTTPS yet, which makes this configuration unsafe for production.

To start the configuration process, connect via SSH to your new MeiliSearch Instance and follow the instructions that appear.

### 2.1. Run the configuration script

Open a terminal and start a new SSH connection with the **Public IPv4** or **your domain name**, using 'admin' as the username.

```bash
ssh admin@<your-ipv4-address>
```

```bash
ssh admin@<your-domain-name>
```

> Write `yes` and press `Enter` to accept the authentication process.

Reviewer: Write yes when asked what?

A script will run automatically, asking for your settings and desired configuration. If you want to run this script again at any time, you can do so by using the following command:

```bash
meilisearch-setup
```

### 3. Enjoy your ready-to-use MeiliSearch instance

Your MeiliSearch Instance is up and running on AWS, and it is ready to be used in **production**.

To check if everything is running smoothly, do another HTTP call to the `/health` route:

```bash
curl -v https://<your-domain-name>/health
```

> Note that this time, we're using HTTPS.

The server should answer with a `204 No content` status code as shown in the example below:

```bash
...
< HTTP/1.1 204 No Content
...
```

> You're all set to use MeiliSearch in production with AWS!

Reviewer: Is our image in a repo somewhere? If not, should we put it out there so people can contribute to it? No matter about that, we should probably finish with something like: if you have any issues with our AWS image please create issues in the XYZ repo.

Author: @eskombro

Reviewer 3: https://github.com/meilisearch/meilisearch-aws

Reviewer 3: Creating an issue into this repo is definitely a good way to help us if there is any trouble 👍

Reviewer 2: Do you mean the AMI itself? If yes, it is stored in our AWS account (for MeilI orga) but the rope that @curquiza linked has all the tools that we use to build it, so it is public and open to contributions

Author: Thanks!
